### PR TITLE
support NeoX-style rotary embedding

### PR DESCRIPTION
**Pull request by wangyems**
### Description

1. The parity is tested under 1e-6.
2. Performance(fp32) w/o rotary embedding by running graph with single op with iobinding.

Before:
Use rotary? False Input shape: torch.Size([1, 32, 1024]) ORT time:  0.09400367736816406 ms
Use rotary? False Input shape: torch.Size([1, 128, 1024]) ORT time:  0.16728639602661133 ms
Use rotary? False Input shape: torch.Size([1, 512, 1024]) ORT time:  0.5242371559143066 ms
Use rotary? False Input shape: torch.Size([1, 1024, 1024]) ORT time:  1.2384796142578125 ms
Use rotary? False Input shape: torch.Size([1, 2048, 1024]) ORT time:  2.367434501647949 ms
Use rotary? False Input shape: torch.Size([8, 32, 1024]) ORT time:  0.23715734481811523 ms
Use rotary? False Input shape: torch.Size([8, 128, 1024]) ORT time:  0.7375001907348633 ms
Use rotary? False Input shape: torch.Size([8, 512, 1024]) ORT time:  2.7140283584594727 ms
Use rotary? False Input shape: torch.Size([8, 1024, 1024]) ORT time:  6.164829730987549 ms
Use rotary? False Input shape: torch.Size([8, 2048, 1024]) ORT time:  15.50304889678955 ms

After:
Use rotary? True Input shape: torch.Size([1, 32, 1024]) ORT time:  0.09786367416381836 ms
Use rotary? True Input shape: torch.Size([1, 128, 1024]) ORT time:  0.19414424896240234 ms
Use rotary? True Input shape: torch.Size([1, 512, 1024]) ORT time:  0.5307626724243164 ms
Use rotary? True Input shape: torch.Size([1, 1024, 1024]) ORT time:  1.1226248741149902 ms
Use rotary? True Input shape: torch.Size([1, 2048, 1024]) ORT time:  2.3289132118225098 ms
Use rotary? True Input shape: torch.Size([8, 32, 1024]) ORT time:  0.2610206604003906 ms
Use rotary? True Input shape: torch.Size([8, 128, 1024]) ORT time:  0.7494926452636719 ms
Use rotary? True Input shape: torch.Size([8, 512, 1024]) ORT time:  2.711203098297119 ms
Use rotary? True Input shape: torch.Size([8, 1024, 1024]) ORT time:  6.184995174407959 ms
Use rotary? True Input shape: torch.Size([8, 2048, 1024]) ORT time:  15.505502223968506 ms

### CODE REVIEW

File: onnxruntime/core/graph/contrib_ops/bert_defs.cc [hunk 1]
```diff
@@ -199,6 +199,10 @@ ONNX_MS_OPERATOR_SET_SCHEMA(
               "(2, batch_size, num_heads, max_sequence_length, head_size)",
               AttributeProto::INT,
               OPTIONAL_VALUE)
+        .Attr("do_rotary",
+              "Whether use rotary position embedding. Default value is 0.",
+              AttributeProto::INT,
+              static_cast<int64_t>(0))
```
Comment (yufenglee) [inline onnxruntime/core/graph/contrib_ops/bert_defs.cc:205]: please add same attribute for QAttention

Comment (wangyems) [reply onnxruntime/core/graph/contrib_ops/bert_defs.cc]: added

File: onnxruntime/contrib_ops/cuda/bert/add_bias_transpose.cu [hunk 2]
```diff
@@ -518,8 +622,35 @@ template <typename T>
...
-    const T* input, const T* biases, T* output, T* qkv_add_bias, const int v_head_size, int total_matrix_count) {
+    const T* input, const T* biases, T* output, T* qkv_add_bias, const int v_head_size, int total_matrix_count,
+    bool do_rotary = false, int original_past_sequence_length = 0) {
   assert(num_heads <= max_threads_per_block);
+
+  if (do_rotary) {
+#ifdef USE_ROCM
+    ORT_THROW("Rotary attention is not supported on ROCm");
+#elif defined(__CUDA_ARCH__) && __CUDA_ARCH__ > 520
+    if (format != 1 && format != 3) {
```
Comment (yufenglee) [inline onnxruntime/contrib_ops/cuda/bert/add_bias_transpose.cu:633]: Why limit it to format 1 and format 3?

Comment (wangyems) [reply onnxruntime/contrib_ops/cuda/bert/add_bias_transpose.cu:633]: it corresponds to gpt2 case with unfused kernel format and cutlass kernel format

Comment (yufenglee) [reply onnxruntime/contrib_ops/cuda/bert/add_bias_transpose.cu:633]: we can still enable fused kernel for the context run.

Comment (wangyems) [reply onnxruntime/contrib_ops/cuda/bert/add_bias_transpose.cu:633]: yeah we can extend the support to all formats later

File: docs/ContribOperators.md [hunk 3]
```diff
@@ -127,6 +127,8 @@ This version of the operator has been available since version 1 of the 'com.micr
 #### Attributes
 
 <dl>
+<dt><tt>do_rotary</tt> : int</dt>
+<dd>Whether use rotary position embedding. Default value is 0.</dd>
```
Comment (justinchuby) [inline docs/ContribOperators.md:131]:
```suggestion
<dd>Whether to use rotary position embedding. Default value is 0.</dd>
```

nit

File: onnxruntime/contrib_ops/cuda/bert/rotary_embedding_util.h [hunk 4]
```diff
@@ -0,0 +1,1292 @@
...
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+#pragma once
```
Comment (yufenglee) [inline onnxruntime/contrib_ops/cuda/bert/rotary_embedding_util.h:19]: we may have a following PR to put utils not specific to another file

**Review by yufenglee — DISMISSED** [review]
:shipit:

**Review by yufenglee — APPROVED** [review]
:shipit:

**Pull request closed without merging**